tests/testsuite/build_plan.rs

```diff
@@ -0,0 +1,182 @@
+                "cwd": "[..]/target/cit/[..]/foo",
+                "deps": [],
+                "display_name": "foo v0.5.0 [..] Target(bin: foo) Profile(dev) Host",
+                "env": "{...}",
+                "links": "{...}",
+                "outputs": "{...}",
+                "program": "rustc"
+            }
+        ]
+    }
+    "#,
+        ),
+    );
+    assert_that(&p.bin("foo"), is_not(existing_file()));
+}
+
+#[test]
+fn cargo_build_plan_single_dep() {
+    let p = project("foo")
+        .file(
+            "Cargo.toml",
+            r#"
+            [package]
+            name = "foo"
+            authors = []
+            version = "0.5.0"
+
+            [dependencies]
+            bar = { path = "bar" }
+        "#,
+        )
+        .file(
+            "src/lib.rs",
+            r#"
+            extern crate bar;
+            pub fn foo() { bar::bar(); }
+
+            #[test]
+            fn test() { foo(); }
+        "#,
+        )
+        .file(
+            "bar/Cargo.toml",
+            r#"
+            [package]
+            name = "bar"
+            version = "0.0.1"
+            authors = []
+        "#,
+        )
+        .file("bar/src/lib.rs", "pub fn bar() {}")
+        .build();
+    assert_that(
+        p.cargo("build").arg("--build-plan"),
+        execs().with_status(0).with_json(
+            r#"
+    {
+        "inputs": [
+            "[..]/foo/Cargo.toml",
+            "[..]/foo/bar/Cargo.toml"
+        ],
+        "module_map": "{...}",
+        "modules": [
+            {
+                "args": "{...}",
+                "cwd": "[..]/target/cit/[..]/foo",
+                "deps": [1],
+                "display_name": "foo v0.5.0 [..] Target(lib) Profile(dev) Host",
+                "env": "{...}",
+                "links": "{...}",
+                "outputs": [
+                    "[..]/foo/target/debug/deps/libfoo-[..].rlib"
+                ],
+                "program": "rustc"
+            },
+            {
+                "args": "{...}",
+                "cwd": "[..]/target/cit/[..]/foo",
+                "deps": [],
+                "display_name": "bar v0.0.1 [..] Target(lib) Profile(dev) Host",
+                "env": "{...}",
+                "links": "{...}",
+                "outputs": [
+                    "[..]/foo/target/debug/deps/libbar-[..].rlib"
+                ],
+                "program": "rustc"
+            }
+        ]
+    }
+    "#,
+        ),
+    );
+}
+
+#[test]
+fn cargo_build_plan_build_script() {
+    let p = project("foo")
+        .file(
+            "Cargo.toml",
+            r#"
+            [project]
+
+            name = "foo"
+            version = "0.5.0"
+            authors = ["wycats@example.com"]
+            build = "build.rs"
+        "#,
+        )
+        .file("src/main.rs", r#"fn main() {}"#)
+        .file("build.rs", r#"fn main() {}"#)
+        .build();
+
+    assert_that(
+        p.cargo("build").arg("--build-plan"),
```

Comment:
Ah one thing I've forgotten when reviewing is that we'll want the `--build-plan` argument to initially be unstable. This can be done by either renaming it to `-Z build-plan` or leaving it as this argument but also requiring `-Z unstable-options` to be passed to enable the argument